### PR TITLE
Add support for managing user profiles

### LINKED ISSUE #769: Keycloak User Profile Configuration
### Problem Statement

Keycloaks Declarative User Profile Feature (feaure = declarative-user-profile) is not available for configuration through keycloak-config-cli.

### Proposed Solution

Add some functionality to enable configuration through keycloak-config-cli.

### Environment

- Keycloak Version: [e.g. 19.0.1]
- keycloak-config-cli Version: [e.g. 1.5.3]
- Java Version: [e.g. 11]

### Additional information

The export/import functionality of keycloak does not contain the information of the user profile component.

Try to focus on the profile information and strip everything else away.

Make use of the admin-api functionality that is used by the admin-ui using JSON.

### Acceptance Criteria

- [ ] As a config cli user I can control the user profile settings

### PR DESCRIPTION
**Pull request by srose**
**What this PR does / why we need it**:
Keycloak export does not contain the User Profile JSON Definiton.
It would be great to have the Profile Definition work with keycloak-config-cli

**Which issue this PR fixes** : 
fixes #769 

**Special notes for your reviewer**:
Want to discuss the solution
Want to aadd some tests.

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

**Comment by srose** [comment]
Would be great to have some feedback...

File: src/test/java/de/adorsys/keycloak/config/service/ImportUserPorfile.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import org.junit.jupiter.api.Order;
+import org.junit.jupiter.api.Test;
+import org.keycloak.representations.idm.RealmRepresentation;
+
+import java.io.IOException;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.Is.is;
+
+public class ImportUserPorfile extends AbstractImportIT {
```
Comment (thomasdarimont) [inline src/test/java/de/adorsys/keycloak/config/service/ImportUserPorfile.java:13]:
```suggestion
public class ImportUserProfile extends AbstractImportIT {
```

File: src/test/java/de/adorsys/keycloak/config/service/ImportUserPorfile.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    private static final String REALM_NAME = "realmWithProfile";
+
+    ImportUserPorfile() {
```
Comment (thomasdarimont) [inline src/test/java/de/adorsys/keycloak/config/service/ImportUserPorfile.java:17]:
```suggestion
    ImportUserProfile() {
```

File: src/main/java/de/adorsys/keycloak/config/repository/UserProfileRepository.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    @Autowired
+    public UserProfileRepository(RealmRepository realmRepository) {
+        this.realmRepository = realmRepository;
+    }
+
+    public void updateUserProfile(String realm, String str) {
+
+        var realmAttributes = realmRepository.get(realm).getAttributesOrEmpty();
+        var userProfileEnabled = Boolean.valueOf(realmAttributes.getOrDefault("userProfileEnabled", "false"));
```
Comment (thomasdarimont) [inline src/main/java/de/adorsys/keycloak/config/repository/UserProfileRepository.java:43]: Extract `userProfileEnabled` to constant.

**Review by thomasdarimont — CHANGES_REQUESTED** [review]
Looks good so far, some minor remarks inline.

File: src/test/resources/import-files/user-federation/00_create_realm_with_federation.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "enabled": true,
-  "realm": "realmWithLdap",
+  "realm": "realmWithProfile",
```
[the remark this answers is not in the document]
Comment (srose) [reply src/test/resources/import-files/user-federation/00_create_realm_with_federation.json:3]: No, i should do that

File: src/main/java/de/adorsys/keycloak/config/repository/UserProfileRepository.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        if (!userProfileEnabled) {
+            return;
+        }
+
+        var userProfileResource = getResource(realm);
+        var profileDefintionHasNotChanged = userProfileResource.getConfiguration().equals(str);
```
[the remark this answers is not in the document]
Comment (srose) [reply src/main/java/de/adorsys/keycloak/config/repository/UserProfileRepository.java:49]: Nice idea, will change that.

**Comment by srose** [comment]
@thomasdarimont please have a look, i am still not sure if its complete. tests could be improved anyhow and i will continue on them.

File: src/main/java/de/adorsys/keycloak/config/repository/UserProfileRepository.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        if (!userProfileEnabled) {
+            userProfileResource.update(null);
+            return;
+        }
+
+        if (userProfileEnabled && userProfileResource == null) {
+            return;
+        }
+
+        String profileResourceConfiguration = Optional.ofNullable(userProfileResource.getConfiguration()).orElse("");
```
Comment (thomasdarimont) [inline src/main/java/de/adorsys/keycloak/config/repository/UserProfileRepository.java:63]:
```suggestion
        var profileResourceConfiguration = Optional.ofNullable(userProfileResource.getConfiguration()).orElse("");
```

**Review by thomasdarimont — CHANGES_REQUESTED** [review]

**Review by thomasdarimont — CHANGES_REQUESTED** [review]
Looks good so far, some remarks:

- Use `var` where appropriate
- Document profile support in https://github.com/adorsys/keycloak-config-cli/blob/main/docs/FEATURES.md
- Fix test failures
- Add minimal profile example config as yaml to https://github.com/adorsys/keycloak-config-cli/tree/main/contrib/example-config

Example config:
```yaml
realm: acme-profile
displayName: "Acme Profile"
enabled: true

attributes:
  userProfileEnabled: "true"

userProfile:
  - name: username
    displayName: "${username}"
    validations:
      length:
        min: 3
        max: 255
      username-prohibited-characters: { }
  - name: email
    displayName: "${email}"
    validations:
      email: { }
      length:
        max: 255
  - name: firstName
    displayName: "${firstName}"
    required:
      roles:
        - user
      scopes: [ ]
    permissions:
      view:
        - user
        - admin
      edit:
        - user
        - admin
    validations:
      length:
        max: 255
      person-name-prohibited-characters: { }
    selector:
      scopes: []
  - name: lastName
    displayName: "${lastName}"
    required:
      roles:
        - user
      scopes: []
    permissions:
      view:
        - user
        - admin
      edit:
        - user
        - admin
    validations:
      length:
        max: 255
      person-name-prohibited-characters: { }
    selector:
      scopes: []
  - name: phoneNumber
    displayName: "${phoneNumber}"
    permissions:
      view:
        - user
        - admin
      edit:
        - user
        - admin
    required:
      roles: [ ]
      scopes:
        - phone
    selector:
      scopes:
        - phone
```

In upcoming versions, we could have dedicated types for profile attributes to enable syntax validation by keycloak-config-cli.

File: src/main/java/de/adorsys/keycloak/config/repository/UserProfileRepository.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            throws JsonProcessingException {
+        if (StringUtils.isEmpty(newPprofileResourceConfiguration)) {
+            newPprofileResourceConfiguration = "";
+        }
+        if (StringUtils.isEmpty(profileResourceConfiguration)) {
+            profileResourceConfiguration = "";
+        }
+
+        var currentValue = mapper.readValue(profileResourceConfiguration, Map.class);
+        var newValue = mapper.readValue(newPprofileResourceConfiguration, Map.class);
```
Comment (thomasdarimont) [inline src/main/java/de/adorsys/keycloak/config/repository/UserProfileRepository.java:86]: I just gave this a try and noticed that the `ObjectMapper` is not provided as a bean if we run the cli tool.
It seems that other places of the code use the Methods from `de.adorsys.keycloak.config.util.JsonUtil` to use a shared object mapper.

Suggestion use:
```
var currentValue = JsonUtil.fromJson(profileResourceConfiguration, Map.class);
var newValue = JsonUtil.fromJson(newPprofileResourceConfiguration, Map.class);
```

and add a method like:
```
    public static <T> T fromJson(String value, Class<T> type) {
        try {
            return objectMapper.readValue(value, type);
        } catch (JsonProcessingException e) {
            throw new ImportProcessingException(e);
        }
    }
```

After that you can remove the objectMapper constructor parameter and field (and unused imports).

**Review by thomasdarimont — CHANGES_REQUESTED** [review]

File: src/main/java/de/adorsys/keycloak/config/model/RealmImport.java [hunk 8]
```diff
@@ -27,12 +27,15 @@
...
 
 import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.List;
 
 @Component
 public class RealmImport extends RealmRepresentation {
     private List<AuthenticationFlowImport> authenticationFlowImports;
 
+    private List<HashMap<String, Object>> userProfile;
```
Comment (thomasdarimont) [inline src/main/java/de/adorsys/keycloak/config/model/RealmImport.java:37]:
```suggestion
    private List<LinkedHashMap<String, Object>> userProfile;
```
The order of entry should be kept as is.

File: src/main/java/de/adorsys/keycloak/config/model/RealmImport.java [hunk 9]
```diff
@@ -49,6 +52,16 @@ public void setAuthenticationFlowImports(List<AuthenticationFlowImport> authenti
         this.authenticationFlowImports = authenticationFlowImports;
     }
 
+    @SuppressWarnings("unused")
+    @JsonSetter("userProfile")
+    public void setUserProfile(List<HashMap<String, Object>> userProfile) {
```
Comment (thomasdarimont) [inline src/main/java/de/adorsys/keycloak/config/model/RealmImport.java:57]:
```suggestion
    public void setUserProfile(List<LinkedHashMap<String, Object>> userProfile) {
```

File: src/main/java/de/adorsys/keycloak/config/model/RealmImport.java [hunk 10]
```diff
@@ -49,6 +52,16 @@ public void setAuthenticationFlowImports(List<AuthenticationFlowImport> authenti
...
+        this.userProfile = userProfile;
+    }
+
+    public List<HashMap<String, Object>> getUserProfile() {
```
Comment (thomasdarimont) [inline src/main/java/de/adorsys/keycloak/config/model/RealmImport.java:61]:
```suggestion
    public List<LinkedHashMap<String, Object>> getUserProfile() {
```

**Review by thomasdarimont — CHANGES_REQUESTED** [review]

File: src/main/java/de/adorsys/keycloak/config/repository/UserProfileRepository.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+    public void updateUserProfile(String realm, Boolean newUserProfileEnabled, String newProfileResourceConfiguration) {
+        var realmAttributes = realmRepository.get(realm).getAttributesOrEmpty();
+
+        var userProfileResource = getResource(realm);
+        if (userProfileResource == null) {
+            logger.error("Cannot retrieve userprofile resource");
+            return;
+        }
+        var profileResourceConfiguration = Optional.ofNullable(userProfileResource.getConfiguration()).orElse("");;
```
Comment (thomasdarimont) [inline src/main/java/de/adorsys/keycloak/config/repository/UserProfileRepository.java:57]:
```suggestion
        var profileResourceConfiguration = Optional.ofNullable(userProfileResource.getConfiguration()).orElse("");
```

**Comment by thomasdarimont** [comment]
@st3v0rr I think this PR is quite complete. Would you mind considering this for the next release of keycloak-config-cli?

**Comment by thomasdarimont** [comment]
@st3v0rr This could be released as a maintenance release 5.3.2

**Comment by st3v0rr** [comment]
Hi @thomasdarimont and @srose,
PR looks good for me I will do a short test and will release it this afternoon.
Due to the new feature 5.4.0 sounds best for me
Thanks for the contribution I didn't find the time to invest that much in the project in the last 2 months :-/
Greets

**Comment by thomasdarimont** [comment]
@st3v0rr Thanks for merging!

We know what it means to maintain an OSS library beside a day job :) We're here to support you :)

**Pull request merged**